controllers/topology/desired_state.go

```diff
@@ -70,11 +73,23 @@ func (r *ClusterReconciler) computeDesiredState(ctx context.Context, s *scope.Sc
 	}
 
 	desiredState.MachineDeployments = map[string]*scope.MachineDeploymentState{}
-	for _, mdTopology := range s.Blueprint.Topology.Workers.MachineDeployments {
-		desiredMachineDeployment, err := computeMachineDeployment(ctx, s, mdTopology)
+	// Sort machine deployment topologies in ascending order of name.
+	// Doing this so that we have a predictable order of rolling out upgrades to machine deployments.
+	mdTopologies := machineDeploymentTopologiesByName(s.Blueprint.Topology.Workers.MachineDeployments)
+	sort.Sort(mdTopologies)
+	upgradeCount := 0
+	allowUpgrade := true
```

Comment:
I wonder if it's easier to first compute all MachineDeployments and afterwards decide which one to upgrade

(I would then also extract the MachineDeployment + MachineDeployment version computation of all MachineDeployments in a computeMachineDeployments func)

Reply:
>What about moving this two info in scope, so we are not required to pass them in and out from the func? This will help in keeping upgrade logic only in computeControlPlane/computeMachineDeployment...
>Let's imagine to have something like scope.desiredState.UpgradeProgress; also maxMachineDeploymentParallelUpgrades could be a method of this struct, thus moving all the infos of the upgrade strategy in one place ...

This sounds good. However, IMO I think `UpgradeProgress` would probably be better places in `scope` rather than `scope.desiredState`. Making `UpgradeProcess` a filed in `ClusterState` would mean it will also show up in `scope.currentState` which could be misleading. 

>I wonder if it's easier to first compute all MachineDeployments and afterwards decide which one to upgrade

I am not sure if thing would make it easier. It would meaning having to go over the computed desired machine deployments and figuring out which of the machine deployments are version changes which include other changes and acting accordingly. It might add some complications that we might want to avoid.  

It would be easier (and also probably cleaner) if computeMachineDeployment gave us the final desired machine deployment. Once we move some of the upgrade progress counter and `maxMachineDeploymentParallelUpgrades` into scope the computeMachineDeployment function should become more cleaner.   

With this, computeMachineDeployment function will also take into account the "upgrade strategy" when calculating the machine deployment.